xml/hwbp_registers.xml

```diff
@@ -3,9 +3,13 @@
     and Debug Mode to prevent untrusted user code from causing entry into Debug
     Mode without the OS's permission.
 
-    In this section XLEN means MXLEN when in M-mode, and DXLEN when in Debug Mode.
-    Note that this makes several of the fields in \RcsrTdataOne move around based
-    on the current execution mode and value of MXLEN.
+    In this section XLEN means MXLEN when in M-mode, and DXLEN when in Debug
+    Mode. On systems where XLEN is writable, this is handled as follows.
+    Fields retain their values regardless of XLEN, which only affects where
+    in the register these fields appear (e.g. \FcsrTdataOneType). Some fields
+    are wider when XLEN is 64 than when it is 32 (e.g.
+    \FcsrTextraThirtytwoSvalue). The high bits in such fields retain their
```

Comment:
So if I’ve set svalue to something with high bits set and then xLen changes, I can’t ever set another breakpoint using scontext since there is no way to clear the upper bits of svalue.  Is that really what we want?

Reply:
If software changes XLEN, it can deal with the complexity of making sure any triggers it set will work with the new XLEN. (Note that you will always be able to access the high bits from the debugger, since DXLEN never changes.) I think this is essential because any change between M-Mode to D-Mode can be an XLEN change, and going and zeroing out a bunch of bits every time that happens sounds like a side effect that could be worse, and there's nothing software can do about it.

Reply:
Ok, two questions then.  Suppose DXLEN=64 and MXLEN can be changed between 32 and 64.  The underlying state will then store 34 bits for svalue.

1. While MXLEN=64, M-mode writes tdata2 such that upper bits of svalue are non-zero.  Then MXLEN is changed to 32.  Then M-mode writes tdata2.  Does the underlying state of svalue[35:16] retain its state or is it set to 0?

2. If svalue[33:16] retains its value, what part of svalue is used when comparing svalue to scontext?  The spec says "match if the low bits of scontext equal svalue".  Does this mean svalue[15:0] or svalue[33:0]?

Reply:
1. I think it makes sense to clear the upper bits if a field is written in 32-bit mode. Cheap to implement, and more likely to do what the users want.
2. In the referred issue, James argued that all bits should still matter. But zeroing out the high bits nicely solves this for the specific case where the trigger was written.

I'll update the PR so the inaccessible bits are zeroed out.

Reply:
Why is DXLEN defined to be the widest supported XLEN instead of just DXLEN=MXLEN?  If MXLEN=SXLEN=UXLEN=32 then why does a debugger need to see the upper 32b that the software being debugged can't see?  That state doesn't affect anything.

Reply:
DXLEN is the widest supported XLEN because it seemed simpler that way. Part of it was me being ornery at the notion that XLEN can change at all. But if XLEN can change every time the hart runs, the debugger needs to discover it every time the hart halts. I'm not convinced it was the right decision, but it seems like a big thing to change right now.

Reply:
I wasn't proposing to change it one business day before the freeze.  I just want to understand the history and if I was missing some important benefit.  It's a bit weird but not the weirdest thing I've seen. 😄

Reply:
More details at https://github.com/riscv/riscv-debug-spec/pull/298